### PR TITLE
ALC-94 basic routing

### PR DESCRIPTION
**Pull request by ledes**
Add basic routing wrapper to redirect to 404 once we have the rules to valid the path
- [x] Add Route wrapper
- [x] Add 404 page
- [x] Add page
- [x] Redirect to 404 if route is different than dashboard
- [ ] Validation of route. Pending

**Note**: I had to remove  `homepage` from package.json as I was having the same issue than @hosainnet

### CODE REVIEW

**Comment by pygeek** [comment]
@ledes Update with latest main to get continuous deployments working.

File: package.json [hunk 1]
```diff
@@ -26,7 +26,6 @@
   "bugs": {
     "url": "https://github.com/alchemistcoin/copper/issues"
   },
-  "homepage": "https://github.com/alchemistcoin/copper#readme",
```
Comment (pygeek) [inline package.json]: This change is causing your conflict with main.

File: src/components/RouteWrapper/index.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+interface Props extends RouteProps {
+  component: React.ComponentType<any>;
+  requiresValidation: boolean;
+}
+const RouteWrapper = ({
+  component: Component,
+  requiresValidation,
+  ...rest
+}: Props) => {
+  const isValid = () => {
```
Comment (pygeek) [inline src/components/RouteWrapper/index.tsx:13]: Consider making `isValid` available for unit testing.

**Comment by ledes** [comment]
> @ledes Update with latest main to get continuous deployments working.

I see .. let me update it  and also the test. Nice work on the setup

File: src/components/RouteWrapper/index.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    // TODO function to validate  if the route and parameters are valid
+    //if valid
+    return true;
+    // else redirect to 404
+  };
+
+  if (requiresValidation && !isValid()) {
```
Comment (pygeek) [inline src/components/RouteWrapper/index.tsx:20]: Consider restructuring things to make logic available to be unit tested (ie: use classes)

**Review by pygeek — CHANGES_REQUESTED** [review]
Add corresponding test files for each file added. You may be able to get away with adding a couple integration tests for 404, though unit test(s) for RouteWrapper may be beneficial once we start building things out.

**Comment by pygeek** [comment]
Consider not deleting the App.tsx test file, instead update it to ensure your functionality works.

**Comment by pygeek** [comment]
I can assist with writing the initial tests if necessary. @ledes

File: package.json (hunk 1 above)
Comment (ledes) [reply package.json]: I think this should be fixed now that is up to date

**Comment by pygeek** [comment]
@ledes Also, was the pull request template visible for you when creating this PR?: https://github.com/alchemistcoin/copper/blob/main/PULL_REQUEST_TEMPLATE.md

**Comment by ledes** [comment]
> I can assist with writing the initial tests if necessary. @ledes

sure! I was going to take a look at what you meant for having classes for those functions

**Comment by pygeek** [comment]
> > I can assist with writing the initial tests if necessary. @ledes
> 
> sure! I was going to take a look at what you meant for having classes for those functions

https://reactjs.org/docs/components-and-props.html#function-and-class-components

**Comment by ledes** [comment]
> @ledes Also, was the pull request template visible for you when creating this PR?: https://github.com/alchemistcoin/copper/blob/main/PULL_REQUEST_TEMPLATE.md

Yes! it was the first timeI saw it, but didn't know you guys wrote tests for every file. One thing more I learned  today

**Comment by pygeek** [comment]
> > @ledes Also, was the pull request template visible for you when creating this PR?: https://github.com/alchemistcoin/copper/blob/main/PULL_REQUEST_TEMPLATE.md
> 
> Yes! it was the first timeI saw it, but didn't know you guys wrote tests for every file. One thing more I learned today

Not necessarily every file needs a test, but we should create a skeleton for tests to be added if adding a new empty tsx file.

**Comment by pygeek** [comment]
@ledes Honestly, it's up to you. If we don't use classes in the other projects that's fine, but we should definitely compose things so that they can be unit tested. I found a good tutorial for both approaches (notice how the functional components define methods inside of an object rather than casting it to a variable—this makes the method available to be executed directly by a test or otherwise): https://dev.to/danielleye/react-class-component-vs-function-component-with-hooks-13dg

Generally, I follow the rule of "if it's testable it's easier to maintain."

**Comment by ledes** [comment]
> @ledes Honestly, it's up to you—we should probably set a standard across our projects or at least this project so we don't have varying conventions in the same repo. If we don't use classes in the other projects that's fine, but we should definitely compose things so that they can be unit tested. I found a good tutorial for both approaches (notice how the functional components define methods inside of an object rather than casting it to a variable—this makes the method available to be executed directly by a test or otherwise): https://dev.to/danielleye/react-class-component-vs-function-component-with-hooks-13dg
> 
> Generally, I follow the rule of "if it's testable it's easier to maintain."

I think it is  a good point. So  just to be on the same page, you are proposing to have isValid in a class component that returns true or false?

**Comment by pygeek** [comment]
@ledes Ping me on Discord. Let's move this discussion offline—possibly have a quick voice chat/screen share if necessary.

**Review by hosainnet — APPROVED** [review]
LGTM

**Comment by pygeek** [comment]
I'm a little confused by this PR based on linked ticket "ALC-89 A user can access a Fair Launch Pool through Copper". Is this being merged in after it's approved or is there additional work that needs to be done for this PR?

It may be a good idea to create a ticket for basic routing, rename your branch to match that so that we can get that merged in and ALC-87 and others can benefit from this change before you continue working on Fair Launch Pool.

**Comment by ledes** [comment]
@pygeek This  pR is just a basic route handling until we have rules to validate the params. Since I didn't have those I decided to just get the basics out and talk about it during stand up
